File: app/src/main/java/me/magnum/melonds/ui/inputsetup/InputSetupViewModel.java
package me.magnum.melonds.ui.inputsetup;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.subjects.BehaviorSubject;
import me.magnum.melonds.model.ControllerConfiguration;
import me.magnum.melonds.model.Input;
import me.magnum.melonds.model.InputConfig;
import me.magnum.melonds.repositories.SettingsRepository;

import java.util.ArrayList;
import java.util.List;

public class InputSetupViewModel extends ViewModel {
    private SettingsRepository settingsRepository;

    private ArrayList<StatefulInputConfig> inputConfigs;
    private BehaviorSubject<List<StatefulInputConfig>> inputConfigsBehaviour;
    private CompositeDisposable disposables;

    public InputSetupViewModel(SettingsRepository settingsRepository) {
        this.settingsRepository = settingsRepository;

        ControllerConfiguration currentConfiguration = settingsRepository.getControllerConfiguration();
        this.inputConfigs = new ArrayList<>();
        for (InputConfig config : currentConfiguration.getConfigList()) {
            this.inputConfigs.add(new StatefulInputConfig(config.clone()));
        }

        this.inputConfigsBehaviour = BehaviorSubject.createDefault((List<StatefulInputConfig>) this.inputConfigs);
        this.disposables = new CompositeDisposable();
    }

    public LiveData<List<StatefulInputConfig>> getInputConfig() {
        final MutableLiveData<List<StatefulInputConfig>> inputConfigLiveData = new MutableLiveData<>();
        Disposable disposable = this.inputConfigsBehaviour.subscribe(new Consumer<List<StatefulInputConfig>>() {
            @Override
            public void accept(List<StatefulInputConfig> statefulInputConfigs) {
                inputConfigLiveData.setValue(statefulInputConfigs);
            }
        });
        this.disposables.add(disposable);

        return inputConfigLiveData;
    }

    public void startUpdatingInputConfig(Input input) {
        for (int i = 0; i < inputConfigs.size(); i++) {
            InputConfig inputConfig = inputConfigs.get(i).getInputConfig();
            if (inputConfig.getInput() == input) {
                inputConfigs.get(i).setIsBeingConfigured(true);
                this.onConfigsChanged();
                break;
            }
        }
    }

    public void stopUpdatingInputConfig(Input input) {
        for (int i = 0; i < inputConfigs.size(); i++) {
            InputConfig inputConfig = inputConfigs.get(i).getInputConfig();
            if (inputConfig.getInput() == input) {
                inputConfigs.get(i).setIsBeingConfigured(false);
                this.onConfigsChanged();
                break;
            }
        }
    }

    public void updateInputConfig(Input input, int key) {
        for (int i = 0; i < inputConfigs.size(); i++) {
            InputConfig inputConfig = inputConfigs.get(i).getInputConfig();
            if (inputConfig.getInput() == input) {
                inputConfig.setKey(key);
                inputConfigs.get(i).setIsBeingConfigured(false);
                this.onConfigsChanged();
                break;
            }
        }
    }

    public void clearInput(Input input) {
        updateInputConfig(input, InputConfig.KEY_NOT_SET);
    }

    private void onConfigsChanged() {
        inputConfigsBehaviour.onNext(this.inputConfigs);

        ControllerConfiguration currentConfiguration = this.buildCurrentControllerConfiguration();
        this.settingsRepository.setControllerConfiguration(currentConfiguration);
    }

    private ControllerConfiguration buildCurrentControllerConfiguration() {
        ArrayList<InputConfig> configs = new ArrayList<>();
        for (StatefulInputConfig statefulConfig : this.inputConfigs) {
            configs.add(statefulConfig.getInputConfig().clone());
        }
        return new ControllerConfiguration(configs);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        this.disposables.dispose();
    }
}

File: app/src/main/java/me/magnum/melonds/ui/inputsetup/InputSetupViewModel.kt
package me.magnum.melonds.ui.inputsetup

import androidx.lifecycle.ViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import me.magnum.melonds.domain.model.ControllerConfiguration
import me.magnum.melonds.domain.model.Input
import me.magnum.melonds.domain.model.InputConfig
import me.magnum.melonds.domain.repositories.SettingsRepository
import me.magnum.melonds.utils.EventSharedFlow
import javax.inject.Inject

@HiltViewModel
class InputSetupViewModel @Inject constructor(private val settingsRepository: SettingsRepository) : ViewModel() {

    private val _inputConfig = MutableStateFlow(settingsRepository.getControllerConfiguration().inputMapper)
    val inputConfiguration = _inputConfig.asStateFlow()

    private val _inputUnderAssignment = MutableStateFlow<Input?>(null)
    val inputUnderAssignment = _inputUnderAssignment.asStateFlow()

    private val _onInputAssignedEvent = EventSharedFlow<Input>()
    val onInputAssignedEvent = _onInputAssignedEvent.asSharedFlow()

    fun startInputAssignment(input: Input) {
        _inputUnderAssignment.value = input
    }

    fun stopInputAssignment() {
        _inputUnderAssignment.value = null
    }

    fun updateInputAssignedKey(key: Int) {
        val inputUnderAssignment = _inputUnderAssignment.value ?: return
        val inputType = InputConfig.Assignment.Key(null, key)
        setInputAssignment(inputUnderAssignment, inputType)
        focusOnNextInput(inputUnderAssignment)
    }

    fun updateInputAssignedAxis(axis: Int, direction: InputConfig.Assignment.Axis.Direction) {
        val inputUnderAssignment = _inputUnderAssignment.value ?: return
        val inputType = InputConfig.Assignment.Axis(null, axis, direction)
        setInputAssignment(inputUnderAssignment, inputType)
        focusOnNextInput(inputUnderAssignment)
    }

    fun clearInputAssignment(input: Input) {
        setInputAssignment(input, InputConfig.Assignment.None)
        _inputUnderAssignment.value = null
    }

    private fun setInputAssignment(input: Input, assignment: InputConfig.Assignment) {
        val inputIndex = _inputConfig.value.indexOfFirst { it.input == input }
        if (inputIndex >= 0) {
            _inputConfig.update { config ->
                config.toMutableList().apply {
                    this[inputIndex] = this[inputIndex].copy(assignment = assignment)
                }.also {
                    onConfigsChanged(it)
                }
            }
        }
        _inputUnderAssignment.value = null
    }

    private fun onConfigsChanged(newConfig: List<InputConfig>) {
        val currentConfiguration = ControllerConfiguration(newConfig)
        settingsRepository.setControllerConfiguration(currentConfiguration)
    }

    private fun focusOnNextInput(currentInput: Input) {
        val currentInputIndex = _inputConfig.value.indexOfFirst { it.input == currentInput }
        val nextInput = _inputConfig.value.getOrNull(currentInputIndex + 1)
        if (nextInput != null) {
            _onInputAssignedEvent.tryEmit(nextInput.input)
        }
    }
}